noria-server/src/controller/sql/mir/join.rs

```diff
@@ -100,5 +113,184 @@ fn pick_join_chains(
+        let (base_c, base_d, join_cd) = make_nodes();
+        let join_abcd = MirNode::new(
+            "c",
+            0,
+            vec![Column::from("aa"), Column::from("ba")],
+            MirNodeType::Join {
+                on_left: vec![Column::from("ab")],
+                on_right: vec![Column::from("bb")],
+                project: vec![Column::from("aa"), Column::from("ba")],
+            },
+            vec![],
+            vec![],
+        );
+
+        node_for_rel.insert("A", base_a);
+        node_for_rel.insert("B", base_b);
+        node_for_rel.insert("C", base_c);
+        node_for_rel.insert("D", base_d);
+        let mut join_chains = Vec::new();
+
+        let (left_chain, second_chain) = pick_join_chains(&"A".to_string(), &"B".to_string(), &mut join_chains, &node_for_rel);
+        match second_chain {
+            Some(right_chain) => {
+                let new_chain = left_chain.merge_chain(right_chain, join_ab.clone());
+                assert!(new_chain.has_table(&"A".to_string()));
+                assert!(new_chain.has_table(&"B".to_string()));
+                assert!(!new_chain.has_table(&"C".to_string()));
+                assert!(!new_chain.has_table(&"D".to_string()));
+                join_chains.push(new_chain);
+            },
+            None => {
+                assert!(false);
+            },
+        };
+
+        let (left_chain, second_chain) = pick_join_chains(&"C".to_string(), &"D".to_string(), &mut join_chains, &node_for_rel);
+        match second_chain {
+            Some(right_chain) => {
+                let new_chain = left_chain.merge_chain(right_chain, join_cd.clone());
+                assert!(new_chain.has_table(&"C".to_string()));
+                assert!(new_chain.has_table(&"D".to_string()));
+                assert!(!new_chain.has_table(&"A".to_string()));
+                assert!(!new_chain.has_table(&"B".to_string()));
+                join_chains.push(new_chain);
+            },
+            None => {
+                assert!(false);
+            },
+        };
+
+        // now join the two pairs together
+        let (left_chain, second_chain) = pick_join_chains(&"D".to_string(), &"B".to_string(), &mut join_chains, &node_for_rel);
+        match second_chain {
+            Some(right_chain) => {
+                let new_chain = left_chain.merge_chain(right_chain, join_abcd.clone());
+                assert!(new_chain.has_table(&"A".to_string()));
+                assert!(new_chain.has_table(&"B".to_string()));
+                assert!(new_chain.has_table(&"C".to_string()));
+                assert!(new_chain.has_table(&"D".to_string()));
+                join_chains.push(new_chain);
+            },
+            None => {
+                assert!(false);
+            },
+        };
+
+        // now query a different combination of them
+        let (left_chain, second_chain) = pick_join_chains(&"A".to_string(), &"C".to_string(), &mut join_chains, &node_for_rel);
```

Comment:
I'm trying to think through whether this test is correct: if I understand correctly, it says "I'm trying to join tables A and C, do we have any join chains for that?", and the result is the chain with A, B, C, and D. But is an ABCD chain really usable for an AC join?`What if the chain (as in this test) first joins A and B and then joins the result with C? The rows in the ABC output aren't necessarily the same as those in an AC join (in particular, there may be more or fewer rows depending on what rows B has).

Now, if the join were ACB, that's different: the chain works because we can just reuse the first two nodes. But `JoinChain` has no notion of table ordering, so we can't tell which one it is.